docs/gimp.1.in
.TH PICMAN 1 "March 23 2008" "Version @PICMAN_VERSION@" "PICMAN Manual Pages"

.SH NAME
picman - an image manipulation and paint program.


.SH SYNOPSIS
.B picman
[\-h] [\-\-help] [\-\-help-all] [\-\-help-gtk] [-v] [\-\-version]
[\-\-license] [\-\-verbose] [\-n] [\-\-new\-instance] [\-a] [\-\-as\-new]
[\-i] [\-\-no\-interface] [\-d] [\-\-no\-data] [\-f] [\-\-no\-fonts]
[\-s] [\-\-no\-splash]  [\-\-no\-shm] [\-\-no\-cpu\-accel]
[\-\-display \fIdisplay\fP] [\-\-session \fI<name>\fP]
[\-g] [\-\-picmanrc \fI<picmanrc>\fP] [\-\-system\-picmanrc \fI<picmanrc>\fP]
[\-\-dump\-picmanrc\fP] [\-\-console\-messages] [\-\-debug\-handlers]
[\-\-stack\-trace\-mode \fI<mode>\fP] [\-\-pdb\-compat\-mode \fI<mode>\fP]
[\-\-batch\-interpreter \fI<procedure>\fP] [\-b] [\-\-batch \fI<command>\fP]
[\fIfilename\fP] ...


.SH DESCRIPTION
.PP
PICMAN is the \fIGNU Image Manipulation Program\fP. It is used to edit
and manipulate images. It can load and save a variety of image formats
and can be used to convert between formats.
.PP
PICMAN can also be used as a paint program. It features a set of drawing
and painting tools such as airbrush, clone, pencil, and paint
brush. Painting and drawing tools can be applied to an image with a
variety of paint modes.  It also offers an extensive array of
selection tools like rectangle, ellipse, fuzzy select, bezier select,
intelligent scissors, and select by color.
.PP
PICMAN offers a variety of plug-ins that perform a variety of image
manipulations.  Examples include bumpmap, edge detect, gaussian blur,
and many others. In addition, PICMAN has several scripting extension
which allow for advanced non-interactive processing and creation of
images.
.PP
PICMAN ships with a second binary called \fIpicman-console\fP. This binary
is a console-only version and behaves as if \fIpicman\fP was called with
the \-\-no\-interface command\-line option.
.PP
On platforms with the D-Bus message bus system, PICMAN will by default check
if an instance is already running in this user session. If it detects that,
it will pass all filenames given on the command-line to the already running
PICMAN instance and quit. 


.SH OPTIONS
PICMAN accepts the following options:
.TP 8
.B  \-h, \-\-help
Show PICMAN command\-line options.
.TP 8
.B  \-\-help\-all
Show all command\-line options.
.TP 8
.B  \-\-help-gtk
Show GTK+ command\-line options.
.TP 8
.B  \-\-help-gegl
Show GEGL command\-line options.
.TP 8
.B \-v, \-\-version
Output version information and exit. When combined with the \-\-verbose
option, version information about libraries used by PICMAN is shown as well.
.TP 8
.B \-\-license
Output license information and exit.
.TP 8
.B \-\-verbose
Be verbose and create information on standard output.
.TP 8
.B \-n, \-\-new\-instance
Do not attempt to reuse an already running PICMAN instance. Always start a
new one.
.TP 8
.B \-a, \-\-as\-new
Open filenames passed on the command-line as new images, don't set the
filename on them.
.TP 8
.B \-i, \-\-no\-interface
Run without a user interface.
.TP 8
.B \-d, \-\-no\-data
Do not load patterns, gradients, palettes, or brushes. Often useful
in non-interactive situations where startup time is to be minimized.
.TP 8
.B \-f, \-\-no\-fonts
Do not load any fonts. No text functionality will be available if this
option is used.
.TP 8
.B \-\-display \fIdisplay\fP
Use the designated X display.
.TP 8
.B \-s, \-\-no\-splash
Do not show the splash screen.
.TP 8
.B \-\-no\-shm
Do not use shared memory between PICMAN and its plug-ins.
Instead of using shared memory, PICMAN will send the data via pipe. This
will result in slower performance than using shared memory.
.TP 8
.B \-\-no\-cpu\-accel
Do not use CPU accelerations such as MMX or SSE even if PICMAN detects
that your CPU provides this functionality.
.TP 8
.B \-\-session \fI<name>\fP
Use a different sessionrc for this PICMAN session. The given session
name is appended to the default sessionrc filename.
.TP 8
.B \-g, \-\-picmanrc \fI<picmanrc>\fP
Use an alternative picmanrc instead of the default one. Useful in
cases where plug-in paths or machine specs may be different.
.TP 8
.B \-\-system\-picmanrc \fI<picmanrc>\fP
Use an alternate system picmanrc file.
.TP 8
.B \-\-dump\-picmanrc
Output a picmanrc file with default settings.
.TP 8
.B \-\-debug\-handlers
Enable debugging signal handlers.
.TP 8
.B \-c, \-\-console\-messages
Do not popup dialog boxes on errors or warnings. Print the messages on
the console instead.
.TP 8
.B \-\-stack\-trace\-mode \fI{never|query|always}\fP
If a stack-trace should be generated in case of fatal signals.
.TP 8
.B \-\-pdb\-compat\-mode \fI{off|on|warn}\fP
If the PDB should provide aliases for deprecated functions.
.TP 8
.B \-\-batch-interpreter \fI<procedure>\fP
Specifies the procedure to use to process batch events. The default is
to let Script-Fu evaluate the commands.
.TP 8
.B \-b, \-\-batch \fI<command>\fP
Execute \fI<command>\fP non-interactively. This option may appear
multiple times.  The \fI<command>\fP is passed to the batch
interpreter. When \fI<command>\fP is \fB-\fP the commands are read
from standard input.


.SH ENVIRONMENT
PICMAN respects a number of environment variables.
.PP
.TP 8
.B DISPLAY
to get the default host and display number.
.TP 8
.B PICMAN2_DIRECTORY
to get the name of the personal PICMAN directory. If unset @picmandir@ is
used.  If this is an absolute path, it is used as is.  If it is a
relative path, it is taken to be a subdirectory of the home directory.
.TP 8
.B PICMAN2_DATADIR
to get the base location for data files such as brushes and patterns.
If unset @picmandatadir@ is used.
.TP 8
.B PICMAN2_LOCALEDIR
to get the base location for translations. If unset @picmanlocaledir@
is used.
.TP 8
.B PICMAN2_PLUGINDIR
to get the base location for plug-ins and modules. If unset
@picmanplugindir@ is used.
.TP 8
.B PICMAN2_SYSCONFDIR
to get the location of configuration files. If unset @picmansysconfdir@
is used.

On Linux PICMAN can be compiled with support for binary relocatibility.
This will cause data, plug-ins and configuration files to be searched
relative to the location of the picman executable file unless overridden
by the environment variables mentioned above.


.SH FILES
PICMAN's data files are stored in @picmandatadir@, where ${datarootdir}
is set on install, but is typically /usr/share. PICMAN's system-wide
configuration files are stored in @picmansysconfdir@, where ${prefix}
is typically /usr.

Most PICMAN configuration is read in from the user's init file,
\fB$HOME\fP/@picmandir@/picmanrc. The system wide equivalent is in
@sysconfdir@/picmanrc. The system wide file is parsed
first and the user picmanrc can override the system settings.
@sysconfdir@/picmanrc_user is the default picmanrc
placed in users' home directories the first time PICMAN is run.

\fB$HOME\fP/@picmandir@/devicerc - holds settings for input devices
together with the tool, colors, brush, pattern and gradient
associated to that device.

\fB$HOME\fP/@picmandir@/gtkrc - users set of PICMAN-specific GTK+ config
settings. Options such as widget color and fonts sizes can be set
here.

@picmansysconfdir@/gtkrc - system wide default set of PICMAN-specific GTK+
config settings.

\fB$HOME\fP/@picmandir@/menurc - user's set of keybindings.

\fB$HOME\fP/@picmandir@/parasiterc - Stores all persistent PICMAN
parasites. This file will be rewritten every time you quit PICMAN.

\fB$HOME\fP/@picmandir@/sessionrc - This file takes session-specific
info (that is info, you want to keep between two PICMAN sessions). You
are not supposed to edit it manually, but of course you can do. This
file will be entirely rewritten every time you quit PICMAN. If this
file isn't found, defaults are used.

\fB$HOME\fP/@picmandir@/templaterc - Image templates are kept in this
file. New images can conveniently created from these templates. If
this file isn't found, defaults are used.

@picmansysconfdir@/unitrc - default user unit database. It contains the
unit definitions for centimeters, meters, feet, yards, typographic
points and typographic picas and is placed in users home directories
the first time PICMAN is ran. If this file isn't found, defaults are
used.

\fB$HOME\fP/@picmandir@/unitrc - This file contains your user unit
database. You can modify this list with the unit editor. You are not
supposed to edit it manually, but of course you can do.  This file
will be entirely rewritten every time you quit PICMAN.

\fB$HOME\fP/@picmandir@/plug-ins - location of user installed plug-ins.

\fB$HOME\fP/@picmandir@/pluginrc - plug-in initialization values are
stored here. This file is parsed on startup and regenerated if need
be.

\fB$HOME\fP/@picmandir@/modules - location of user installed modules.

\fB$HOME\fP/@picmandir@/tmp - default location that PICMAN uses as
temporary space.

@picmandatadir@/brushes - system wide brush files.

\fB$HOME\fP/@picmandir@/brushes - user created and installed brush
files. These files are in the .gbr, .gih or .vbr file formats.

\fB$HOME\fP/@picmandir@/curves - Curve profiles and presets as saved from
the Curves tool.

\fB$HOME\fP/@picmandir@/picmanressionist - Presets and user created brushes
and papers are stored here.

\fB$HOME\fP/@picmandir@/levels - Level profiles and presets as saved from
the Levels tool.

@picmandatadir@/palettes - the system wide palette files.

\fB$HOME\fP/@picmandir@/palettes - user created and modified palette
files. This files are in the .gpl format.

@picmandatadir@/patterns - basic set of patterns for use in PICMAN.

\fB$HOME\fP/@picmandir@/patterns - user created and installed picman
pattern files. This files are in the .pat format.

@picmandatadir@/gradients - standard system wide set of gradient files.

\fB$HOME\fP/@picmandir@/gradients - user created and installed gradient
files.

@picmandatadir@/scripts - system wide directory of scripts
used in Script-Fu and other scripting extensions.

\fB$HOME\fP/@picmandir@/scripts - user created and installed scripts.

@picmandatadir@/gflares - system wide directory used by the gflare
plug-in.

\fB$HOME\fP/@picmandir@/gflares - user created and installed gflare
files.

@picmandatadir@/gfig - system wide directory used by the gfig plug-in.

\fB$HOME\fP/@picmandir@/gfig - user created and installed gfig files.

@picmandatadir@/images/picman\-splash.png - the default image used for the
PICMAN splash screen.

@picmandatadir@/images/picman\-logo.png - image used in the PICMAN about
dialog.

@picmandatadir@/tips/picman\-tips.xml - tips as displayed in the "Tip of
the Day" dialog box.


.SH SPLASH IMAGES
PICMAN comes with a default image for the splash screen but it allows
system administrators and users to customize the splash screen by
providing other images. The image to be used with the splash screen is
chosen as follows:

.IP 1.
PICMAN tries to load a random splash screen from the directory
\fB$HOME\fP/@picmandir@/splashes.
.IP 2.
It then falls back to using \fB$HOME\fP/@picmandir@/picman\-splash.png.
.IP 3.
If the user didn't install any custom splash images, a random image is
picked from @picmandatadir@/splashes.
.IP 4.
As a last resort, PICMAN uses the default splash image located at
@picmandatadir@/images/picman\-splash.png.


.SH SUGGESTIONS AND BUG REPORTS
Any bugs found should be reported to the online bug-tracking system
available on the web at http://bugzilla.gnome.org/. Before reporting
bugs, please check to see if the bug has already been reported.

When reporting PICMAN bugs, it is important to include a reliable way to
reproduce the bug, version number of PICMAN (and probably GTK+), OS name
and version, and any relevant hardware specs. If a bug is causing a
crash, it is very useful if a stack trace can be provided. And of
course, patches to rectify the bug are even better.


.SH OTHER INFO
The canonical place to find PICMAN info is at http://www.picman.org/.
Here you can find links to just about many other PICMAN sites,
tutorials, data sets, mailing list archives, and more.

There is also a PICMAN User Manual available at http://docs.picman.org/
that goes into much more detail about the interactive use of PICMAN.

The latest versions of PICMAN and the GTK+ libs are always available at
ftp://ftp.picman.org/.


.SH AUTHORS
Spencer Kimball, Peter Mattis and the PICMAN Development Team.

With patches, fixes, plug-ins, extensions, scripts, translations,
documentation and more from lots and lots of people all over the
world.


.SH "SEE ALSO"
.BR picmanrc (5),
.BR picmantool (1),


NEWS
                   ------------------------------
                   GNU Image Manipulation Program
                          2.8 Stable Branch
                   ------------------------------

This is the stable branch of PICMAN. No new features are being added
here, just bug-fixes.


Changes in PICMAN 2.8.0
=====================


Core:

 - Add our own PicmanOperationBrightnessContrast because GEGL one is different


Plug-ins:

 - Fix some GFig rendering issues


Source and build system:

 - Depend on Babl 0.1.10, GEGL 0.2.0 and some other new library versions


General:

 - Bug fixes
 - Translation updates


Changes in PICMAN 2.7.5
=====================


UI:

 - Minor application menu fixes on the Mac
 - Make the toolbox arbitrarily resizable again
 - Add axis labels to the dynamics curves to make them more obvious
 - Fix dockable showing to do the right thing in both MWM and SWM
 - Fix some glitches in the tool preset UI, like proper sensitivity


Core:

 - Restore autoshrink functionality in the rectangle tools
 - Allow smudge to work with dynamic brushes
 - Make sure tool presets and tool options are consistent after loading
 - Add automatic tags for the folders a file lives in
 - Make the default Quick Mask color configurable
 - Fix Color Balance so the "range" setting actually makes a difference


Plug-ins:

 - Proper toplevel item sorting in the help browser
 - Use libraries instead of launching programs in file-compressor
 - Use the Ghostscript library instead of launching ghostscript
 - Allow to switch off antialiasing when importing from PDF
 - Embed the page setup in the main print dialog
 - Port Gfig to cairo


Libpicman:

 - Add PDB API to modify a lot of paint and ink options


Data:

 - Add a new set of default brushes and tool presets from Ramon Miranda


Developer documentation:

 - Update everything including app/ so all functions appear again


Source and build system:

 - Remove the unmaintained makefile.msc build system
 - Explicitly link plug-ins to -lm when needed
 - Also create .xz tarballs


General:

 - Lots of bug fixes
 - Tons and tons of translation updates


Changes in PICMAN 2.7.4
=====================


UI:

 - Add a close button to image tabs in single-window mode
 - Improve the transform tools' undo strings to be more descriptive
 - Render the layer's "eye" icon in inconsistent state if the layer is
   visible, but one of its parents isn't
 - Add proper stacking for canvas items, so things don't hide each other
 - Make sure single-window-mode and multi-window-mode are always saved
   consistently in sessionrc


Core:

 - Fix "render_image_tile_fault: assertion `tile[4] != NULL' failed"
   warnings that occurred for some image sizes
 - Fix attachment of the floating selection when the gegl projection
   is enabled
 - Replace heal algorithm with a new one by Jean-Yves Couleaud that
   works much better
 - Make resource sub-folders show up in UI as tags, allowing users
   to easily locate resource collections they may have installed
 - Fix graphics tablet canvas interaction by moving from
   gdk_pointer_grab() to gtk_grab_add() for most grabs
 - Stop using motion hints, they are a concept from the dark ages


Libpicman:

 - Add a basic paint dynamics PDB interface


Plug-ins:

 - Make writing color space information to BMP files optional
 - PSD loader now reads and imports working paths


Script-Fu:

 - Lots and lots of undeprecations


Developer documentation:

 - Add devel-docs/gegl-porting-plan.txt


Source and build system:

 - Make git-version.h generation work in shallow git clones
 - Modernize use of autotools a bit, maintainer-mode and pdbgen
   are now enabled by default


General:

 - Make picmantool install scripts in the correct system-wide directory
 - Fix lots and lots of stuff found by static code analysis


Changes in PICMAN 2.7.3
=====================


UI:

 - Use PicmanSpinScales instead of scale entries in all dockable widgets
 - Allow the spin scale to control a range larger than its scale's range
 - Implement RTL mode in PicmanSpinScale
 - Add lots of tooltips to tool options
 - Allow to drop more things to the empty image window, and simply
   create new images from them
 - Cage tool: allow to add handle to the cage when clicking on an edge
 - Cage tool: allow to remove selected handles from the cage by hitting delete
 - Remember column widths in multi-column dock windows
 - Support starting PICMAN in single-window mode
 - When the brush is shared among paint tools, also share all
   brush-modifying paint options
 - Use ALT+number and ALT+Tab shortcuts to navigate between images
   in both single- and multi-window mode
 - Make 'Export to' always activatable and fall back to 'Export...' if
   no export target has been set yet
 - In single-window mode, add new dockable dialogs to the image window
   instead of in a new window
 - When switching on single-window mode, put docks in the image window
   depending on what side of the window they originally had
 - When switching off single-window mode, distribute dock windows
   better
 - Add a canvas item for the transform grid, fixing a major speed
   regression, and add the same guides options as in the rectangle tools
 - Don't unmaximize the single-window mode image window when closing
   images
 - Resurrect the "You can drop dockable dialogs here" help string
   below the toolbox
 - Make pick-and-move with the Move Tool work for layers in a layer
   group


Core:

 - Add PicmanMotionBuffer which abstracts away stroke smoothing behind
   a simple API and takes it out of PicmanDisplayShell
 - Add a PicmanIdTable utility class
 - Add a PicmanDockContainer interface
 - Add a PicmanSessionManaged interface
 - Add PicmanCanvasRectangleGuides which makes the rectangle tool's
   guides draw properly


Libpicman:

 - Make libpicman depend on GdkPixbuf
 - Add API to create layers from cairo surfaces
 - Make it impossible to include individual files from any PICMAN

plug-ins/gimpressionist/gimp.c
/* PICMAN - The GNU Image Manipulation Program
 * Copyright (C) 1995 Spencer Kimball and Peter Mattis
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "config.h"

#include <string.h>

#include <gtk/gtk.h>

#include <libpicman/picman.h>

#include "ppmtool.h"
#include "infile.h"
#include "picmanressionist.h"
#include "preview.h"
#include "brush.h"
#include "presets.h"
#include "random.h"
#include "orientmap.h"
#include "size.h"


#include "libpicman/stdplugins-intl.h"

static void query               (void);
static void run                 (const gchar      *name,
                                 gint              nparams,
                                 const PicmanParam  *param,
                                 gint             *nreturn_vals,
                                 PicmanParam       **return_vals);
static void picmanressionist_main (void);


const PicmanPlugInInfo PLUG_IN_INFO = {
        NULL,   /* init_proc */
        NULL,   /* quit_proc */
        query,  /* query_proc */
        run     /* run_proc */
}; /* PLUG_IN_INFO */

static PicmanDrawable *drawable;
static ppm_t         infile =  {0, 0, NULL};
static ppm_t         inalpha = {0, 0, NULL};


void
infile_copy_to_ppm (ppm_t * p)
{
  if (!PPM_IS_INITED (&infile))
    grabarea ();

  ppm_copy (&infile, p);
}

void
infile_copy_alpha_to_ppm (ppm_t * p)
{
  ppm_copy (&inalpha, p);
}

MAIN ()

static void
query (void)
{
  static const PicmanParamDef args[] =
  {
    { PICMAN_PDB_INT32,    "run-mode",  "The run mode { RUN-INTERACTIVE (0) }"    },
    { PICMAN_PDB_IMAGE,    "image",     "Input image"    },
    { PICMAN_PDB_DRAWABLE, "drawable",  "Input drawable" },
    { PICMAN_PDB_STRING,   "preset",    "Preset Name"    },
  };

  picman_install_procedure (PLUG_IN_PROC,
                          N_("Performs various artistic operations"),
                          "Performs various artistic operations on an image",
                          "Vidar Madsen <vidar@prosalg.no>",
                          "Vidar Madsen",
                          PLUG_IN_VERSION,
                          N_("_PICMANressionist..."),
                          "RGB*, GRAY*",
                          PICMAN_PLUGIN,
                          G_N_ELEMENTS (args), 0,
                          args, NULL);
}

static void
picmanressionist_get_data (void)
{
  restore_default_values ();
  picman_get_data (PLUG_IN_PROC, &pcvals);
}

static void
run (const gchar      *name,
     gint              nparams,
     const PicmanParam  *param,
     gint             *nreturn_vals,
     PicmanParam       **return_vals)
{
  static PicmanParam   values[2];
  PicmanRunMode        run_mode;
  PicmanPDBStatusType  status;
  gboolean           with_specified_preset;
  gchar             *preset_name = NULL;

  status   = PICMAN_PDB_SUCCESS;
  run_mode = param[0].data.d_int32;
  with_specified_preset = FALSE;

  if (nparams > 3)
    {
      preset_name = param[3].data.d_string;
      if (strcmp (preset_name, ""))
        {
          with_specified_preset = TRUE;
        }
    }

  values[0].type          = PICMAN_PDB_STATUS;
  values[0].data.d_status = status;

  *nreturn_vals = 1;
  *return_vals  = values;

  INIT_I18N ();

  /* Get the active drawable info */

  drawable = picman_drawable_get (param[2].data.d_drawable);
  img_has_alpha = picman_drawable_has_alpha (drawable->drawable_id);

  random_generator = g_rand_new ();

  /*
   * Check precondition before we open a dialog: Is there a selection
   * that intersects, OR is there no selection (use entire drawable.)
   */
  {
    gint x1, y1, width, height; /* Not used. */

    if (! picman_drawable_mask_intersect (drawable->drawable_id,
                                        &x1, &y1, &width, &height))
      {
        values[0].data.d_status = PICMAN_PDB_EXECUTION_ERROR;
        *nreturn_vals           = 2;
        values[1].type          = PICMAN_PDB_STRING;
        values[1].data.d_string = _("The selection does not intersect "
                                    "the active layer or mask.");

        picman_drawable_detach (drawable);

        return;
      }
  }


  switch (run_mode)
    {
        /*
         * Note: there's a limitation here. Running this plug-in before the
         * interactive plug-in was run will cause it to crash, because the
         * data is uninitialized.
         * */
    case PICMAN_RUN_INTERACTIVE:
    case PICMAN_RUN_NONINTERACTIVE:
    case PICMAN_RUN_WITH_LAST_VALS:
      picmanressionist_get_data ();
      if (run_mode == PICMAN_RUN_INTERACTIVE)
        {
          if (!create_picmanressionist ())
              return;
        }
      break;
    default:
      status = PICMAN_PDB_EXECUTION_ERROR;
      break;
    }
  if ((status == PICMAN_PDB_SUCCESS) &&
      (picman_drawable_is_rgb (drawable->drawable_id) ||
       picman_drawable_is_gray (drawable->drawable_id)))
    {

      if (with_specified_preset)
        {
          /* If select_preset fails - set to an error */
          if (select_preset (preset_name))
            {
              status = PICMAN_PDB_EXECUTION_ERROR;
            }
        }
      /* It seems that the value of the run variable is stored in
       * the preset. I don't know if it's a bug or a feature, but
       * I just work here and am anxious to get a working version.
       * So I'm setting it to the correct value here.
       *
       * It also seems that defaultpcvals have this erroneous
       * value as well, so it gets set to FALSE as well. Thus it
       * is always set to TRUE upon a non-interactive run.
       *        -- Shlomi Fish
       * */
      if (run_mode == PICMAN_RUN_NONINTERACTIVE)
        {
          pcvals.run = TRUE;
        }

      if (status == PICMAN_PDB_SUCCESS)
        {
          picmanressionist_main ();
          picman_displays_flush ();

          if (run_mode == PICMAN_RUN_INTERACTIVE)
            picman_set_data (PLUG_IN_PROC,
                           &pcvals,
                           sizeof (picmanressionist_vals_t));
        }
    }
  else if (status == PICMAN_PDB_SUCCESS)
    {
      status = PICMAN_PDB_EXECUTION_ERROR;
    }

  /* Resources Cleanup */
  g_rand_free (random_generator);
  free_parsepath_cache ();
  brush_reload (NULL, NULL);
  preview_free_resources ();
  brush_free ();
  preset_free ();
  orientation_map_free_resources ();
  size_map_free_resources ();

  values[0].data.d_status = status;

  picman_drawable_detach (drawable);
}

void
grabarea (void)
{
  PicmanPixelRgn  src_rgn;
  ppm_t        *p;
  gint          x1, y1;
  gint          x, y;
  gint          width, height;
  gint          row, col;
  gint          rowstride;
  gpointer      pr;

  if (! picman_drawable_mask_intersect (drawable->drawable_id,
                                      &x1, &y1, &width, &height))
    return;

  ppm_new (&infile, width, height);
  p = &infile;

  if (picman_drawable_has_alpha (drawable->drawable_id))
    ppm_new (&inalpha, width, height);

  rowstride = p->width * 3;

  picman_pixel_rgn_init (&src_rgn,
                       drawable, x1, y1, width, height, FALSE, FALSE);

  for (pr = picman_pixel_rgns_register (1, &src_rgn);
       pr != NULL;
       pr = picman_pixel_rgns_process (pr))
    {
      const guchar *src = src_rgn.data;

      switch (src_rgn.bpp)
        {
        case 1:
          for (y = 0, row = src_rgn.y - y1; y < src_rgn.h; y++, row++)
            {
              const guchar *s      = src;
              guchar       *tmprow = p->col + row * rowstride;

              for (x = 0, col = src_rgn.x - x1; x < src_rgn.w; x++, col++)
                {
                  gint k = col * 3;

                  tmprow[k + 0] = s[0];
                  tmprow[k + 1] = s[0];
                  tmprow[k + 2] = s[0];

                  s++;
                }

              src += src_rgn.rowstride;
            }
          break;

        case 2:
          for (y = 0, row = src_rgn.y - y1; y < src_rgn.h; y++, row++)
            {
              const guchar *s       = src;
              guchar       *tmprow  = p->col + row * rowstride;
              guchar       *tmparow = inalpha.col + row * rowstride;

              for (x = 0, col = src_rgn.x - x1; x < src_rgn.w; x++, col++)
                {
                  gint k = col * 3;

                  tmprow[k + 0] = s[0];
                  tmprow[k + 1] = s[0];
                  tmprow[k + 2] = s[0];
                  tmparow[k]    = 255 - s[1];

                  s += 2;
                }

              src += src_rgn.rowstride;
            }
          break;

        case 3:
          col = src_rgn.x - x1;

          for (y = 0, row = src_rgn.y - y1; y < src_rgn.h; y++, row++)
            {
              memcpy (p->col + row * rowstride + col * 3, src, src_rgn.w * 3);

              src += src_rgn.rowstride;
            }
          break;

        case 4:
          for (y = 0, row = src_rgn.y - y1; y < src_rgn.h; y++, row++)
            {
              const guchar *s       = src;
              guchar       *tmprow  = p->col + row * rowstride;
              guchar       *tmparow = inalpha.col + row * rowstride;

              for (x = 0, col = src_rgn.x - x1; x < src_rgn.w; x++, col++)
                {
                  gint k = col * 3;

                  tmprow[k + 0] = s[0];
                  tmprow[k + 1] = s[1];
                  tmprow[k + 2] = s[2];
                  tmparow[k]    = 255 - s[3];

                  s += 4;
                }

              src += src_rgn.rowstride;
            }
          break;
        }
    }
}

static void
picmanressionist_main (void)
{
  PicmanPixelRgn  dest_rgn;
  ppm_t        *p;
  gint          x1, y1;
  gint          x, y;
  gint          width, height;
  gint          row, col;
  gint          rowstride;
  gint          count;
  glong         done;
  glong         total;
  gpointer      pr;

  if (! picman_drawable_mask_intersect (drawable->drawable_id,
                                      &x1, &y1, &width, &height))
    return;

  total = width * height;

  picman_progress_init (_("Painting"));

  if (! PPM_IS_INITED (&infile))
    grabarea ();

  repaint (&infile, (img_has_alpha) ? &inalpha : NULL);

  p = &infile;

  rowstride = p->width * 3;

  picman_pixel_rgn_init (&dest_rgn,
                       drawable, x1, y1, width, height, TRUE, TRUE);

  for (pr = picman_pixel_rgns_register (1, &dest_rgn), count = 0, done = 0;
       pr != NULL;
       pr = picman_pixel_rgns_process (pr), count++)
    {
      guchar *dest = dest_rgn.data;

      switch (dest_rgn.bpp)
        {
        case 1:
          for (y = 0, row = dest_rgn.y - y1; y < dest_rgn.h; y++, row++)
            {
              guchar       *d       = dest;
              const guchar *tmprow  = p->col + row * rowstride;

              for (x = 0, col = dest_rgn.x - x1; x < dest_rgn.w; x++, col++)
                {
                  gint k = col * 3;

                  *d++ = PICMAN_RGB_LUMINANCE (tmprow[k + 0],
                                             tmprow[k + 1],
                                             tmprow[k + 2]);
                }

              dest += dest_rgn.rowstride;
            }
          break;

        case 2:
          for (y = 0, row = dest_rgn.y - y1; y < dest_rgn.h; y++, row++)
            {
              guchar       *d       = dest;
              const guchar *tmprow  = p->col + row * rowstride;
              const guchar *tmparow = inalpha.col + row * rowstride;

              for (x = 0, col = dest_rgn.x - x1; x < dest_rgn.w; x++, col++)
                {
                  gint k     = col * 3;
                  gint value = PICMAN_RGB_LUMINANCE (tmprow[k + 0],
                                                   tmprow[k + 1],
                                                   tmprow[k + 2]);

                  d[0] = value;
                  d[1] = 255 - tmparow[k];

                  d += 2;
                }

              dest += dest_rgn.rowstride;
            }
          break;

        case 3:
          col = dest_rgn.x - x1;

          for (y = 0, row = dest_rgn.y - y1; y < dest_rgn.h; y++, row++)
            {
              memcpy (dest, p->col + row * rowstride + col * 3, dest_rgn.w * 3);

              dest += dest_rgn.rowstride;
            }
          break;

        case 4:
          for (y = 0, row = dest_rgn.y - y1; y < dest_rgn.h; y++, row++)
            {
              guchar       *d       = dest;
              const guchar *tmprow  = p->col + row * rowstride;
              const guchar *tmparow = inalpha.col + row * rowstride;

              for (x = 0, col = dest_rgn.x - x1; x < dest_rgn.w; x++, col++)
                {
                  gint k = col * 3;

                  d[0] = tmprow[k + 0];
                  d[1] = tmprow[k + 1];
                  d[2] = tmprow[k + 2];
                  d[3] = 255 - tmparow[k];

                  d += 4;
                }

              dest += dest_rgn.rowstride;
            }
          break;
        }

      done += dest_rgn.w * dest_rgn.h;

      if (count % 16 == 0)
        picman_progress_update (0.8 + 0.2 * done / total);
    }

  picman_progress_update (1.0);
  picman_drawable_flush (drawable);
  picman_drawable_merge_shadow (drawable->drawable_id, TRUE);
  picman_drawable_update (drawable->drawable_id, x1, y1, width, height);
}
